### PR TITLE
Feature/add lead popup

### PR DESCRIPTION
**Pull request by AdityaKhatri**
- Addresses https://github.com/the-deep/client/issues/1850

## Changes

* Add new lead add/edit form

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [ ] translations

### CODE REVIEW

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+    const sortedPriority = useMemo(() => (
+        leadOptions?.priority?.sort((a, b) => compareNumber(a.key, b.key))
+    ), [leadOptions?.priority]);
+
+    return (
+        <div className={_cs(styles.leadEditForm, className)}>
+            {pending && <PendingMessage />}
+            <SelectInput
+                label="Project"
```
Comment (samshara) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx:114]: add `FIXME` for use of translations

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    count: number;
+}
+
+export interface EmmEntity {
+    name: string;
+}
+
+export interface Lead {
+    // TODO: Handle case where assignee can be multiple
+    assignee: number;
```
Comment (samshara) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts:37]: any reason why multiple assignees cannot be assigned right now ?

**Review by samshara — APPROVED** [review]

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/ConfidentialityInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    } = props;
+
+    const modifiedValue = value === 'confidential';
+
+    const handleCheckboxChange = useCallback((newVal: boolean) => {
+        if (newVal) {
+            onChange('confidential', name);
+        } else {
+            onChange('unprotected', name);
+        }
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/ConfidentialityInput/index.tsx]: let's use a ternary operator

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/EmmStats/index.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    className?: string;
+    emmEntities?: EmmEntity[];
+    emmTriggers?: EmmTrigger[];
+}
+
+function EmmStatsModal(props: Props) {
+    const {
+        className,
+        emmEntities = [],
+        emmTriggers = [],
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/EmmStats/index.tsx]: Do we need to do this?

File: src/newViews/Tagging/Sources/LeadEditModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        }),
+        onSuccess: () => {
+            onLeadSaveSuccess();
+        },
+        onFailure: (response, ctx) => {
+            if (response.value.errors) {
+                setError(transformErrorToToggleFormError(schema, ctx, response.value.errors));
+            }
+        },
+        failureHeader: 'Leads',
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/index.tsx]:
```suggestion
        failureHeader: 'Lead',
```

File: src/newViews/Tagging/Sources/LeadEditModal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        pending: leadSavePending,
+        trigger: triggerLeadSave,
+    } = useLazyRequest<Lead, PartialFormType>({
+        url: leadId ? `server://v2/leads/${leadId}/` : 'server://v2/leads/',
+        method: leadId ? 'PATCH' : 'POST',
+        body: ctx => ({
+            ...ctx,
+            authors: ctx.authors ?? [],
+            emmTriggers: ctx.emmTriggers ?? [],
+            emmEntities: ctx.emmEntities ?? [],
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/index.tsx]: on need to do this. form can handle this

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            sourceType: [requiredCondition],
+            title: [requiredStringCondition],
+        };
+        if (value?.sourceType === 'website') {
+            baseSchema = {
+                ...baseSchema,
+                url: [requiredCondition, urlCondition],
+                website: [requiredCondition],
+                emmEntities: [],
+                emmTriggers: [],
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts]: maybe server needs these values for any type.

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    hasEmmLeads: boolean;
+    emmEntities?: EmmEntityOption[];
+    emmRiskFactors?: EmmEntityOption[];
+    emmKeywords?: EmmEntityOption[];
+}
+
+export const schema:FormSchema = {
+    fields: (value): FormSchemaFields => {
+        let baseSchema: FormSchemaFields = {
+            assignee: [requiredCondition],
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts:104]: no including id?

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+                    className={styles.input}
+                    name="authors"
+                    value={value.authors}
+                    onChange={setFieldValue}
+                    options={authorOrganizationOptions ?? leadOptions?.organizations}
+                    onOptionsChange={setAuthorOrganizationOptions}
+                    disabled={pending}
+                    label="Authoring Organizations"
+                    hint={isTruthyString(value?.authorRaw) && `Previous organization: ${value?.authorRaw}`}
+                    // FIXME: Talk with @tnagorra regarding appropriate use of error
```
Comment (tnagorra) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx]: let's talk

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/EmmStats/index.tsx (hunk 4 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/EmmStats/index.tsx]: yes

File: src/newViews/Tagging/Sources/LeadEditModal/index.tsx (hunk 6 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/index.tsx]: how?

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts (hunk 2 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/schema.ts:37]: It is the requirement to set assignee as single select, but in old type I found it as number | number[].
Lets leave it as is for now, and fix it if we run into issues later.

**Review by subinasr — APPROVED** [review]

**Review by tnagorra — APPROVED** [review]

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx (hunk 1 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx:114]: Done for the whole file

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    ] = useState<BasicOrganization[] | undefined | null>();
+
+    const {
+        pending,
+        response: leadOptions,
+    } = useRequest<LeadOptions>({
+        method: 'POST',
+        url: 'server://lead-options/',
+        body: optionsRequestBody,
+        failureHeader: 'Lead Options',
```
Comment (subinasr) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx:102]: shouldn't we refresh the source table on every added source?

**Review by subinasr — CHANGES_REQUESTED** [review]

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx (hunk 10 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/index.tsx:102]: Yes we should, but now its a bit difficult. Need to discuss this. I've added a todo for now.

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/ConfidentialityInput/index.tsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import React, { useCallback } from 'react';
+import {
+    Checkbox,
+    CheckboxProps,
+} from '@the-deep/deep-ui';
+
+interface Props<N extends string> extends Omit<CheckboxProps<N>, 'value' | 'onChange'> {
+    value?: string;
```
Comment (frozenhelium) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/ConfidentialityInput/index.tsx]:
```suggestion
    value?: 'confidential' | 'unprotected';
```

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        }
+
+        .nested-row {
+            display: flex;
+            flex-basis: 50%;
+
+            .nested-input {
+                flex-grow: 1;
+            }
+        }
```
Comment (frozenhelium) [inline src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss:22]: optional: we can just reuse the `row` and `input` class here

File: src/newViews/Tagging/Sources/LeadEditModal/index.tsx [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        onClose,
+        projectId,
+        leadId,
+        onLeadSaveSuccess,
+    } = props;
+
+    const [initialValue, setInitialValue] = useState<PartialFormType>(() => ({
+        project: projectId,
+        sourceType: 'website',
+        priority: 100,
```
Comment (frozenhelium) [inline src/newViews/Tagging/Sources/LeadEditModal/index.tsx:46]: What does the `priority` of `100` mean?

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss (hunk 12 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss:22]: There is a bit of a difference, one grows while one has constant flex-basis of 50%.

File: src/newViews/Tagging/Sources/LeadEditModal/index.tsx (hunk 13 above)
Comment (AdityaKhatri) [reply src/newViews/Tagging/Sources/LeadEditModal/index.tsx:46]: The lowest priority we have right now.

File: src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss (hunk 12 above)
Comment (frozenhelium) [reply src/newViews/Tagging/Sources/LeadEditModal/LeadEditForm/styles.scss:22]: @AdityaKhatri yep, we can do something like this,
```scss
.row {
    display: flex;
    flex-basis: 100%;

    .row {  // only occupy 50% if row is nested
        flex-basis: 50%;
    }
}
```

**Pull request merged**